crates/esthri/src/ops/sync.rs

```diff
@@ -894,6 +943,96 @@ where
+    ///
+    /// # Arguments
+    ///
+    /// * `s3` - S3 client
+    /// * `source` - S3 prefix
+    /// * `destination` - local directory to sync to
+    /// * `glob_filter` - An (optional) slice of filters that specify whether files
+    ///                   should be included or not. These are processed in order,
+    ///                   with the first matching filter determining whether the
+    ///                   file will be included or excluded. If not supplied, then
+    ///                   all files will be synced.
+    #[logfn(err = "ERROR")]
+    pub async fn sync<'a, T>(
+        s3: &'a T,
+        source: &'a S3PathParam,
+        destination: &'a S3PathParam,
+        filters: &'a [GlobFilter],
+        opts: SharedSyncOptParams,
+    ) -> Result<impl Stream<Item = Result<SyncedFile>> + 'a>
+    where
+        T: S3 + Sync + Send + Clone,
+    {
+        let (bucket, key, path) = match (source, destination) {
+            (S3PathParam::Bucket { bucket, key }, S3PathParam::Local { path }) => {
+                info!(
+                    "sync-down, local directory: {}, bucket: {}, key: {}",
+                    path.display(),
+                    bucket,
+                    key
+                );
+                (bucket, key, path)
+            }
+            _ => {
+                warn!("sync streaming is only implemented for s3 to local");
+                return Err(Error::SyncStreamingNotImplemented);
+            }
+        };
+
+        Ok(sync_remote_to_local(s3, bucket, key, path.to_str().unwrap(), filters, opts).await)
+    }
+
+    async fn sync_remote_to_local<'a, T>(
+        s3: &'a T,
+        bucket: &'a str,
+        key: &'a str,
+        directory: &'a str,
+        filters: &'a [GlobFilter],
+        opts: SharedSyncOptParams,
+    ) -> impl Stream<Item = Result<SyncedFile>> + 'a
+    where
+        T: S3 + Sync + Send + Clone,
+    {
+        let task_count = Config::global().concurrent_sync_tasks();
+        let object_listing = flattened_object_listing(
+            s3,
+            bucket,
+            key,
+            Path::new(directory),
+            filters,
+            opts.transparent_compression,
+        )
+        .map_ok(|(path, _key, s3_metadata)| (path, s3_metadata));
+        let cmd = if opts.transparent_compression {
+            SyncCmd::DownCompressed
+        } else {
+            SyncCmd::Down
+        };
+
+        let etag_stream = translate_paths(object_listing, cmd).buffer_unordered(task_count);
+        let sync_tasks = remote_to_local_sync_tasks(
+            s3.clone(),
+            bucket.into(),
+            key.into(),
+            directory.into(),
+            etag_stream,
+            opts,
+        );
+
+        Box::pin(
+            sync_tasks
+                .buffer_unordered(task_count)
+                .take_while(|_task| future::ready(true)), // always go through the whole stream
+        )
```

Comment:
I think just this should work 

```suggestion
        Box::pin(sync_tasks.buffer_unordered(task_count))
```